financial-services-accelerator/components/org.wso2.financial.services.accelerator.consent.mgt.extensions/src/main/java/org/wso2/financial/services/accelerator/consent/mgt/extensions/authorize/impl/DefaultConsentPersistStep.java

```diff
@@ -141,64 +153,102 @@ public void consentPersist(ConsentPersistData consentPersistData, ConsentResourc
+        Map<String, ConsumerAccountDTO> accountNameToObject = new HashMap<>();
+        if (consumerAccounts != null) {
+            for (ConsumerAccountDTO consumerAccount: consumerAccounts) {
+                accountNameToObject.put(consumerAccount.getDisplayName(), consumerAccount);
             }
+        }
+
+        // Set of all accounts to map
+        Set<AccountDTO> allAccountsToMap = new HashSet<>();
 
-            //Check whether account Ids are strings
-            JSONArray accountIds = persistPayload.getJSONArray(ConsentExtensionConstants.ACCOUNT_IDS);
-            for (Object account : accountIds) {
-                if (!(account instanceof String)) {
-                    log.error(ConsentAuthorizeConstants.ACCOUNT_ID_FORMAT_ERROR);
-                    throw new ConsentException(ResponseStatus.BAD_REQUEST,
-                            ConsentAuthorizeConstants.ACCOUNT_ID_FORMAT_ERROR);
+        // Append all consent initiated accounts
+        if (initiatedAccountsForConsent != null) {
+            allAccountsToMap.addAll(initiatedAccountsForConsent);
+        }
+
+        // Append all permission initiated accounts
+        if (permissions != null) {
+            Set<AccountDTO> allAccountsSet = new HashSet<>();
+            for (PermissionDTO permission: permissions) {
+                List<AccountDTO> initiatedAccounts = permission.getInitiatedAccounts();
+                if (initiatedAccounts != null) {
+                    allAccountsSet.addAll(initiatedAccounts);
                 }
-                if (((String) account).isEmpty()) {
-                    if (isApproved) {
-                        log.error(ConsentAuthorizeConstants.ACCOUNT_ID_NOT_FOUND_ERROR);
-                        throw new ConsentException(ResponseStatus.BAD_REQUEST,
-                                ConsentAuthorizeConstants.ACCOUNT_ID_NOT_FOUND_ERROR);
-                    } else {
-                        account = "n/a";
+            }
+            allAccountsToMap.addAll(allAccountsSet);
+        }
+
+        // Append all selected consumer accounts
+        Boolean allowMultipleAccounts = consentData.getAllowMultipleAccounts();
+        allowMultipleAccounts = allowMultipleAccounts != null && allowMultipleAccounts;
+        boolean foundOneAccount = false;
+        JSONObject accountPermissionParameters =
+                persistPayload.optJSONObject(ConsentAuthorizeConstants.REQUEST_ACCOUNT_PERMISSION_PARAMETERS);
+        if (accountPermissionParameters != null) {
+            for (String key: accountPermissionParameters.keySet()) {
+                if (key.contains("accounts")) {
+
+                    // allowMultipleAccounts validation
+                    if (foundOneAccount && !allowMultipleAccounts) {
+                        throw new IllegalStateException("Found multiple account selections when only one is allowed");
+                    }
+
+                    foundOneAccount = true;
```

Comment:
Confused on the usage of the foundOneAccount parameter. Can you please check as discussed offline